fhirflat/ingest.py

```diff
@@ -165,6 +165,38 @@ def create_dict_wide(
                         result[key].append(snippet[key])
                     else:
                         result[key] = [result[key], snippet[key]]
+
+                # Keys that were not previously in the result still need to be added
+                remaining_keys = set(snippet.keys()) ^ duplicate_keys
+                if remaining_keys:
+                    key_length = max(len(result[k]) for k in duplicate_keys)
+                    empty_list = [None] * (key_length - 1)
+                    for key in remaining_keys:
+                        result[key] = [*empty_list, snippet[key]]
+
+                # Check for existing keys that might need to be extended
+                snippet_keys = list(snippet.keys())
+                result_groups = group_keys(result.keys())
+                for k_list in result_groups.values():
+                    if set(snippet_keys).issubset(set(k_list)):
+                        relevant_result = {
+                            k: (
+                                [result[k]]
+                                if not isinstance(result[k], list)
+                                else result[k]
+                            )
+                            for k in k_list
+                        }
+                        all_vals_same_length = (
+                            len(set(map(len, relevant_result.values()))) == 1
+                        )
+                        if not all_vals_same_length:
+                            target_length = max(map(len, relevant_result.values()))
+                            for k, v in relevant_result.items():
+                                if len(v) < target_length:
+                                    result[k] = relevant_result[k] + [None] * (
```

Comment:
This is specifically for input data scenarios where there are lists of data linked together. The scenario being tested is for diagnosis; in FHIR, there is a `diagnosis` attribute that follows this kind of structure:
```
diagnosis: [
              {
                 condition: {dengue, ....},
                 use: {main diagnosis, ....}
              },
              {
                 condition: {cancer, ....},
                 use: {secondary diagnosis, ....}
              },
              {...},...
           ]
```
This is a 'backbone element', i.e. a list of multiple linked structures. It's really hard to expand this out to FHIRflat because you'd end up with a column called diagnosis.condition with a list of conditions, and a column called diagnosis.use with a list of conditions, and only the ordering in the list links the condition with the use case which would be a nightmare to search for, hence the *_dense columns.

We still however have to use a flat-like structure for the mapping file, and make use of the fact that the condition&use will be added at the same time, making the list ordered, to enable the construction of the Backbone element which then gets dumped into a *_dense (in this case `diagnosis_dense`) column. This relies on every list going into the backbone element being of the same length. However there can be scenarios where either a `use` isn't provided for a `condition`, or free text is provided in the coding columns without a code+system (to put in `text` rather than `display` within a codeable concept). In this case to preserve the list ordering & length, when items are added to parts of the backbone element lists but not all of them, the unmodified lists need padding with None to make sure the next item that gets added doesn't get linked incorrectly. E.g. if one row adds the free text 'malaria' to diagnosis.condition.text and the next row of the mapping file adds the full codeable concept "dengue", code=1234, system="snomed". We want to ensure that the 1234 code doesn't get linked to "malaria" rather than "dengue".

Reply:
I realise this ^ is a messy stream of thought explanation, I'll try and explain it more logically tomorrow!

Reply:
Yes - I get that in this case there would be right padding with [malaria, None, None] and [dengue, 1234, snomed]. But there could be a case where there is no free text and a code like [None, 5432, None]- then right padding would only create [5432, None, None]?

Reply:
Oh I see what you mean - no, the padding is added at the point where a new element is appended to the list, iteratively, rather than only at the end.
See the test `test_create_dict_one_to_one_dense_freetext` - the first set of params shows left padding, the second and last show right padding. It will also add padding in the middle if appropriate.